tools/audio/extraction/tuning.py

```diff
@@ -0,0 +1,212 @@
+    "C7",  "DF7",  "D7",  "EF7",  "E7",  "F7",  "GF7",    "G7",    "AF7", "A7", "BF7", "B7",
+    "C8",  "DF8",  "D8",  "EF8",  "E8",  "F8",  "GF8",    "G8",    "AF8", "A8", "BF8", "B8",
+    "C9",  "DF9",  "D9",  "EF9",  "E9",  "F9",  "GF9",    "G9",    "AF9", "A9", "BF9", "B9",
+    "C10", "DF10", "D10", "EF10", "E10", "F10",
+                                                                                "BFNEG1", "BNEG1",
+    "C0",  "DF0",  "D0",  "EF0",  "E0",  "F0",  "GF0",    "G0",    "AF0",
+)
+
+# Floats that are encountered in extraction but cannot be resolved to a match.
+BAD_FLOATS = [0x3E7319E3]
+
+def note_z64_to_midi(note : int) -> int:
+    """
+    Convert a z64 note number to MIDI note number.
+
+    Middle C is 39 in z64, while it is 60 in MIDI.
+    We want MIDI note numbers to store in the extracted sample files (aiff or wav)
+    """
+    return (21 + note) % 128
+
+def recalc_tuning(rate : int, note : str) -> float:
+    return f32(f32(rate / 32000.0) * u32_to_f32(g_pitch_frequencies[pitch_names.index(note)]))
+
+def rate_from_tuning(tuning : float) -> Tuple[Tuple[str,int]]:
+    """
+    Decompose a tuning value into a pair (samplerate, basenote) that round-trips when ran through `recalc_tuning`
+    """
+    matches : List[Tuple[str,int]] = []
+    diffs : List[Tuple[int, Tuple[str,int]]] = []
+
+    tuning_bits : int = f32_to_u32(tuning)
+
+    def test_value(note_val : int, nominal_rate : int, freq : float):
+        if nominal_rate > 48000:
+            # reject samplerate if too high
+            return
+
+        # recalc tuning and compare to original
+
+        tuning2 : float = f32(f32(nominal_rate / 32000.0) * freq)
+
+        diff : int = abs(f32_to_u32(tuning2) - tuning_bits)
+
+        if diff == 0:
+            matches.append((pitch_names[note_val], nominal_rate))
+        else:
+            diffs.append((diff, (pitch_names[note_val], nominal_rate)))
+
+    # search gPitchFrequencies LUT one by one. We don't exit as soon as a match is found as in general this procedure
+    # only recovers the correct (rate,note) pair up to multiples of 2, to get the final value we want to select the
+    # "best" of these pairs by an essentially arbitrary ranking (cf `rank_rates_notes`)
+    for note_val,freq_bits in enumerate(g_pitch_frequencies):
+        freq : float = u32_to_f32(freq_bits)
+
+        # compute the "nominal" samplerate for a given basenote by R = 32000 * (t / f)
+        nominal_rate : int = int(f32(tuning / freq) * 32000.0)
+
+        # test nominal value and +/-1
+        test_value(note_val, nominal_rate,     freq)
+        test_value(note_val, nominal_rate + 1, freq)
+        test_value(note_val, nominal_rate - 1, freq)
+
+    if len(matches) != 0:
+        return tuple(matches)
+
+    # no matches found... check if we expected this, otherwise flag it for special handling
+    assert tuning_bits in BAD_FLOATS , f"0x{tuning_bits:08X}"
```

Comment:
Just curious, when does this occur?

Reply:
This is a can of worms, starting here
https://discord.com/channels/688807550715560050/688851451014938642/1091563920529047712
to
https://discord.com/channels/688807550715560050/688851451014938642/1099140701285003386

the

> just 1 off-by-1

tharo mentions there must be this float in BAD_FLOATS

Reply:
Thanks, what song(s) does it occur in? I did some spelunking and it seems the "bad float" is only in SAMPLE_0_90 which is used in INST_67 of soundfont 0, but I can't find where that instrument is used

Reply:
Four layers in sequence 0 use SF0_INST_67, for implementing the following sfx
````
LAYER_0B1D -> CHAN_0B08 -> NA_SE_PL_ARROW_CHARGE_FIRE
LAYER_0B1D -> CHAN_5E0A -> NA_SE_SY_SET_FIRE_ARROW

LAYER_0B57 -> CHAN_0B37 -> NA_SE_PL_ARROW_CHARGE_ICE
LAYER_0B57 -> CHAN_0B5D -> NA_SE_PL_ARROW_CHARGE_LIGHT

LAYER_14D6 -> CHAN_14CA -> NA_SE_IT_EXPLOSION_FRAME

LAYER_1BEF -> CHAN_15F3 -> NA_SE_IT_FLAME
LAYER_1BEF -> CHAN_1BE9 -> NA_SE_EV_FLAME_IGNITION
````

Reply:
oh thanks, I only looked for SF0_INST_67 in extracted/, I didn't realize there were some checked in at assets/audio/sequences too